src/main/scala/io/iohk/ethereum/vm/DataWord.scala

```diff
@@ -79,10 +88,56 @@ class DataWord private (private val n: BigInt) extends Ordered[DataWord] {
 
   def *(that: DataWord): DataWord = DataWord(this.n * that.n)
 
-  def /(that: DataWord): DataWord = DataWord(this.n / that.n)
+  def /(that: DataWord): DataWord = zeroCheck(that) { DataWord(this.n / that.n) }
+
+  def sdiv(that: DataWord): DataWord = zeroCheck(that) {
+    DataWord(toUnsignedBigInt(this.signedN / that.signedN))
+  }
+
+  def mod(that: DataWord): DataWord = zeroCheck(that) { DataWord(this.n mod that.n) }
+
+  def smod(that: DataWord): DataWord = zeroCheck(that) {
+    DataWord(this.signedN.signum * (this.signedN.abs mod that.signedN.abs))
+  }
+
+  def addmod(that: DataWord, modulus: DataWord): DataWord = zeroCheck(modulus) {
+    DataWord((this.n + that.n) mod modulus.n)
+  }
+
+  def mulmod(that: DataWord, modulus: DataWord): DataWord = zeroCheck(modulus) {
+    DataWord((this.n * that.n) mod modulus.n)
+  }
 
   def **(that: DataWord): DataWord = DataWord(this.n.modPow(that.n, Modulus))
 
+  def signExtend(that: DataWord): DataWord = {
+    val cnt: Byte = that.n.toByte
+    if (cnt < 0 || cnt > 31) {
+      this
+    } else {
+      val mask: Int = if (this.signedN.testBit((cnt * 8) + 7)) 0xFF else 0x00
+      val leftFill: Array[Byte] = Array.fill(Size - cnt - 1)(mask.toByte)
+      val bs = ByteString(leftFill) ++ this.bytes.takeRight(cnt + 1)
+      DataWord(bs)
```

Comment:
Nothing wrong with this, but I would I take a different approach:

```scala
val negative = n.testBit(cnt * 8 + 7)
val mask = (BigInt(1) << ((cnt + 1) * 8)) - 1 // == BigInt(2).pow((cnt + 1) * 8) - 1
val newN = if (negative) n | (MaxValue ^ mask) else n & mask
new DataWord(newN)
```

I would expect this to be a tad faster (though not 100% sure)